### ISSUE TITLE
Transport client doesn't handle DNS -> IP binding changing

### ISSUE THREAD

**Issue by mikemccand** [elastic/elasticsearch#21424]
Today, when you instantiate the transport client, it resolves DNS -> IP on init and uses that IP forever.

If your DNS binding changes, the transport client will keep using the old IP, forcing you to know when to re-instantiate it to get the new DNS binding.

Maybe transport client should more frequently resolve, e.g. per request, or every N seconds?  This is possibly dangerous, though, in the face of flaky/slow DNS lookups, and it could add a sizable latency.

This is similar to https://github.com/elastic/elasticsearch/issues/16412 for unicast zen ping.

**Comment by s1monw** [comment]
I think all we need to do here is to add the ability to notify the user that we disconnected from a node for some reason. Then the user can just go and take some action based on the event adding back a new transport address resolving the DNS name again. I really think that would solve the issue here in a generic way without much overhead. We already did that in the http client @javanna correct me if I am wrong.
